cpp/dolfinx/fem/assembler.h

```diff
@@ -587,22 +610,43 @@ void set_diagonal(auto set_fn, std::span<const std::int32_t> rows,
 /// @param[in] set_fn The function for setting values to a matrix.
 /// @param[in] V The function space for the rows and columns of the
 /// matrix. It is used to extract only the Dirichlet boundary conditions
-/// that are define on V or subspaces of V.
+/// that are defined on V or subspaces of V.
 /// @param[in] bcs The Dirichlet boundary conditions.
 /// @param[in] diagonal Value to add to the diagonal for rows with a
 /// boundary condition applied.
+/// @param[in] unassembled Whether the matrix is in unassembled format
 template <dolfinx::scalar T, std::floating_point U>
 void set_diagonal(
     auto set_fn, const FunctionSpace<U>& V,
     const std::vector<std::reference_wrapper<const DirichletBC<T, U>>>& bcs,
-    T diagonal = 1.0)
+    T diagonal = 1.0, bool unassembled = false)
 {
   for (auto& bc : bcs)
   {
     if (V.contains(*bc.get().function_space()))
     {
       const auto [dofs, range] = bc.get().dof_indices();
-      set_diagonal(set_fn, dofs.first(range), diagonal);
+      if (unassembled)
+      {
+        // We need to insert on ghost indices too; since MATIS is not
+        // designed to support the INSERT_VALUES stage, we scale ALL
+        // the diagonal values we insert into the matrix
+        // TODO: move scaling factor computation to DirichletBC or IndexMap?
+        auto adjlist = bc.get().function_space()->dofmap()->index_map->index_to_dest_ranks();
+        const auto off = adjlist.offsets();
+        std::vector<std::int32_t> number_of_sharing_ranks(off.size());
+        for (size_t i = 0; i < off.size() - 1; ++i)
+            number_of_sharing_ranks[i] = off[i+1] - off[i] + 1;
```

Comment:
Ideally, this setup code should be moved to `IndexMap`
I was thinking of implementing it lazily, with a `mutable` class member, but I see you don't use `mutable` in the dolfinx. Would it be something you would accept?

Reply:
Hi Stefano,
I know I suggested this function on discourse. However, as I now see the use-case, what I think we should do is to 1. create an `la::Vector` on the index map.
2. Fill vector with 1 (including ghosts)
3. scatter reverse add

that would give you an array of the number of shared procs per dof

Reply:
This could probably be a vector we compute once inside the DirichletBC class. What do you think @garth-wells ?

Reply:
I’m without a computer until Sunday. @IgorBaratta could you do me a solid and make a simple example for Stefano?:)

Reply:
> I’m without a computer until Sunday. @IgorBaratta could you do me a solid and make a simple example for Stefano?:)

Sure, I'll post an example as soon as I can today.

Reply:
No rush, I can wait for a proper working implementation

Reply:
We need a way to divide the insertion of 1 on the diagonal by the number of processes that has the dof, as MATIS cannot use insert, and has to add the values. I.e. if a DirichletBC dof is shared between N processes, we assign 1/N on each process.

I suggest we use a vector, stored in the DirichletBC to keep track of this, a illustrated with the snippet above.

Reply:
@jorgensd summarized it properly. If you want to keep the same workflow for assembled and unassembled matrices, the counting vector is the only way to realize it. I think such information should belong to `IndexMap` instead of `DirichletBC`, but I'm fine either way.

Reply:
@stefanozampini  @garth-wells I've made an attempt at this at: https://github.com/FEniCS/dolfinx/tree/dokken/suggested-matis
I'll have a go at installing PETSc main locally and run this, to see if I can get the demo to run:)

Reply:
A likely good approach is consistent handling of the LHS and RHS, i.e. use accumulate rather than set for the bc on the RHS. We did this is legacy DOLFIN. I'd need to look closely at the code for how this might work with the latest design.